File: go-sdk/kestra_api_client/cases_api.go
package kestra_api_client

import (
	"context"
	"encoding/json"
	"os"
)

// CasesAPI covers /api/v1/{tenant}/cases.
type CasesAPI struct {
	baseAPI
}

// CreateCaseFromTask creates a case from a task run, or attaches the triggering
// execution to a matching already-open case. It backs the CreateCase plugin task,
// and is a check-then-act rather than an atomic operation: concurrent runs of the
// same task can each create their own case.
func (a *CasesAPI) CreateCaseFromTask(ctx context.Context, tenant string, request interface{}) (map[string]interface{}, error) {
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", "from-task"), request, nil)
}

// ---- search / counts ----

// SearchCases returns a page of cases matching the given filters.
func (a *CasesAPI) SearchCases(ctx context.Context, tenant string, page, size *int, sort []string, filters []SearchFilter, dateFilter *string) (*PagedResultsCase, error) {
	params := buildQueryParams("page", page, "size", size, "dateFilter", dateFilter)
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[*PagedResultsCase](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", "search"), nil, params)
}

// CaseCounts returns the number of cases per status matching the given filters.
func (a *CasesAPI) CaseCounts(ctx context.Context, tenant string, filters []SearchFilter, dateFilter *string) (map[string]int64, error) {
	params := buildQueryParams("dateFilter", dateFilter)
	appendFilterParams(params, filters)
	return doJSON[map[string]int64](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", "counts"), nil, params)
}

// CaseAssignees lists the distinct assignees among cases matching the given filters.
func (a *CasesAPI) CaseAssignees(ctx context.Context, tenant string, filters []SearchFilter, dateFilter *string) (*PagedResultsCaseSubjectRef, error) {
	params := buildQueryParams("dateFilter", dateFilter)
	appendFilterParams(params, filters)
	return doJSON[*PagedResultsCaseSubjectRef](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", "assignees"), nil, params)
}

// ---- CRUD ----

// CreateCase creates a new case.
func (a *CasesAPI) CreateCase(ctx context.Context, tenant string, request CaseCreateRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases"), request, nil)
}

// GetCase retrieves a case by id. Set allowDeleted to also return a soft-deleted case.
func (a *CasesAPI) GetCase(ctx context.Context, tenant, id string, allowDeleted *bool) (*Case, error) {
	params := buildQueryParams("allowDeleted", allowDeleted)
	return doJSON[*Case](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", id), nil, params)
}

// UpdateCase updates a case's mutable fields.
func (a *CasesAPI) UpdateCase(ctx context.Context, tenant, id string, request CaseUpdateRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "PUT", tenantPath(tenant, "cases", id), request, nil)
}

// DeleteCase soft-deletes a case.
func (a *CasesAPI) DeleteCase(ctx context.Context, tenant, id string) error {
	return a.doVoidJSON(ctx, "DELETE", tenantPath(tenant, "cases", id), nil, nil)
}

// DeleteCasesByIds deletes several cases at once, returning the number deleted.
func (a *CasesAPI) DeleteCasesByIds(ctx context.Context, tenant string, caseIds []string) (*BulkResponse, error) {
	return doJSON[*BulkResponse](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", "by-ids", "delete"), caseIds, nil)
}

// DeleteCasesByQuery deletes every case matching the given filters, returning the number deleted.
func (a *CasesAPI) DeleteCasesByQuery(ctx context.Context, tenant string, filters []SearchFilter, dateFilter *string) (*BulkResponse, error) {
	params := buildQueryParams("dateFilter", dateFilter)
	appendFilterParams(params, filters)
	return doJSON[*BulkResponse](&a.baseAPI, ctx, "DELETE", tenantPath(tenant, "cases", "by-query"), nil, params)
}

// ---- lifecycle ----

// AcknowledgeCasesByIds acknowledges several cases at once, returning the number acknowledged.
func (a *CasesAPI) AcknowledgeCasesByIds(ctx context.Context, tenant string, caseIds []string) (*BulkResponse, error) {
	return doJSON[*BulkResponse](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", "by-ids", "acknowledge"), caseIds, nil)
}

// AcknowledgeCase acknowledges a single case.
func (a *CasesAPI) AcknowledgeCase(ctx context.Context, tenant, id string) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "acknowledge"), nil, nil)
}

// ResolveCase resolves a case with a reason and optional note.
func (a *CasesAPI) ResolveCase(ctx context.Context, tenant, id string, request CaseResolveRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "resolve"), request, nil)
}

// CancelCase cancels a case. The request body is optional.
func (a *CasesAPI) CancelCase(ctx context.Context, tenant, id string, request *CaseCancelRequest) (*Case, error) {
	var body interface{}
	if request != nil {
		body = request
	}
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "cancel"), body, nil)
}

// ChangeCaseStatus sets a case's status, optionally with a resolution.
func (a *CasesAPI) ChangeCaseStatus(ctx context.Context, tenant, id string, request CaseStatusRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "status"), request, nil)
}

// AssignCase sets a case's assignees and/or watchers.
func (a *CasesAPI) AssignCase(ctx context.Context, tenant, id string, request CaseAssignRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "assign"), request, nil)
}

// FollowCase adds the current user as a watcher of a case.
func (a *CasesAPI) FollowCase(ctx context.Context, tenant, id string) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "follow"), nil, nil)
}

// UnfollowCase removes the current user as a watcher of a case.
func (a *CasesAPI) UnfollowCase(ctx context.Context, tenant, id string) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "unfollow"), nil, nil)
}

// ---- events / comments / attachments ----

// CaseEvents returns a page of a case's events (comments and lifecycle events).
func (a *CasesAPI) CaseEvents(ctx context.Context, tenant, id string, page, size *int) (*PagedResultsCaseEvent, error) {
	params := buildQueryParams("page", page, "size", size)
	return doJSON[*PagedResultsCaseEvent](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", id, "events"), nil, params)
}

// AddCaseComment posts a comment on a case, with an optional markdown body and an
// optional file attachment. Pass a nil filePath to comment without an attachment.
func (a *CasesAPI) AddCaseComment(ctx context.Context, tenant, id, body string, filePath *string) (*CaseEvent, error) {
	formParams := map[string]interface{}{}
	if body != "" {
		formParams["body"] = body
	}
	if filePath != nil {
		f, err := os.Open(*filePath)
		if err != nil {
			return nil, err
		}
		defer f.Close()
		formParams["files"] = f
	}

	resp, err := a.doMultipartJSON(ctx, "POST", tenantPath(tenant, "cases", id, "comments"), nil, formParams)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var event CaseEvent
	if err := json.NewDecoder(resp.Body).Decode(&event); err != nil {
		return nil, err
	}
	return &event, nil
}

// DownloadCaseAttachment downloads a comment attachment's raw bytes.
func (a *CasesAPI) DownloadCaseAttachment(ctx context.Context, tenant, id, attachmentId string) ([]byte, error) {
	return a.doDownloadBytes(ctx, "GET", tenantPath(tenant, "cases", id, "attachments", attachmentId), nil, nil)
}

// ---- linked executions ----

// CaseExecutions returns a page of a case's linked executions.
func (a *CasesAPI) CaseExecutions(ctx context.Context, tenant, id string, page, size *int) (*PagedResultsCaseExecution, error) {
	params := buildQueryParams("page", page, "size", size)
	return doJSON[*PagedResultsCaseExecution](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", id, "executions"), nil, params)
}

// LinkCaseExecutions links executions to a case by id.
func (a *CasesAPI) LinkCaseExecutions(ctx context.Context, tenant, id string, request CaseLinkExecutionsRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "executions"), request, nil)
}

// LinkCaseExecutionsByQuery links every execution matching the given filters to a case.
func (a *CasesAPI) LinkCaseExecutionsByQuery(ctx context.Context, tenant, id string, request CaseLinkExecutionsByQueryRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "executions", "by-query"), request, nil)
}

// UnlinkCaseExecution removes a linked execution from a case.
func (a *CasesAPI) UnlinkCaseExecution(ctx context.Context, tenant, id, executionId string) error {
	return a.doVoidJSON(ctx, "DELETE", tenantPath(tenant, "cases", id, "executions", executionId), nil, nil)
}

// ---- auto-attach ----

// EnableCaseAutoAttach enables auto-attaching future matching executions to a case.
func (a *CasesAPI) EnableCaseAutoAttach(ctx context.Context, tenant, id string, request CaseAutoAttachRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "auto-attach"), request, nil)
}

// DisableCaseAutoAttach disables auto-attaching future executions to a case.
func (a *CasesAPI) DisableCaseAutoAttach(ctx context.Context, tenant, id string) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "DELETE", tenantPath(tenant, "cases", id, "auto-attach"), nil, nil)
}

// ---- creation from executions ----

// CreateCaseFromExecutions creates a case and links the given executions to it.
func (a *CasesAPI) CreateCaseFromExecutions(ctx context.Context, tenant string, request CaseFromExecutionsRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", "from-executions"), request, nil)
}

// CreateCaseFromExecutionsByQuery creates a case and links every execution matching the filters.
func (a *CasesAPI) CreateCaseFromExecutionsByQuery(ctx context.Context, tenant string, request CaseFromExecutionsByQueryRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", "from-executions", "by-query"), request, nil)
}

// CasesByExecutions returns, per execution id, the summaries of cases linked to it.
func (a *CasesAPI) CasesByExecutions(ctx context.Context, tenant string, executionIds []string) (map[string][]CaseSummary, error) {
	return doJSON[map[string][]CaseSummary](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", "by-executions"), executionIds, nil)
}

// ---- assets ----

// CaseAssets lists a case's linked assets (explicit and auto-detected).
func (a *CasesAPI) CaseAssets(ctx context.Context, tenant, id string) (*PagedResultsCaseAsset, error) {
	return doJSON[*PagedResultsCaseAsset](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", id, "assets"), nil, nil)
}

// AttachCaseAsset explicitly links an asset to a case.
func (a *CasesAPI) AttachCaseAsset(ctx context.Context, tenant, id string, request CaseAttachAssetRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "assets"), request, nil)
}

// DetachCaseAsset unlinks an asset from a case.
func (a *CasesAPI) DetachCaseAsset(ctx context.Context, tenant, id, assetId string) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "DELETE", tenantPath(tenant, "cases", id, "assets", assetId), nil, nil)
}

// CasesByAsset lists the cases linked to an asset (explicit and auto-detected).
func (a *CasesAPI) CasesByAsset(ctx context.Context, tenant, assetId string) (*PagedResultsCaseSummary, error) {
	return doJSON[*PagedResultsCaseSummary](&a.baseAPI, ctx, "GET", tenantPath(tenant, "cases", "by-asset", assetId), nil, nil)
}

// ---- actions ----

// AttachCaseAction attaches a flow as a runnable action on a case.
func (a *CasesAPI) AttachCaseAction(ctx context.Context, tenant, id string, request CaseAttachActionRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "actions"), request, nil)
}

// UpdateCaseAction replaces an attached action's label or target flow.
func (a *CasesAPI) UpdateCaseAction(ctx context.Context, tenant, id, namespace, flowId string, request CaseAttachActionRequest) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "PUT", tenantPath(tenant, "cases", id, "actions", namespace, flowId), request, nil)
}

// DetachCaseAction detaches a flow action from a case.
func (a *CasesAPI) DetachCaseAction(ctx context.Context, tenant, id, namespace, flowId string) (*Case, error) {
	return doJSON[*Case](&a.baseAPI, ctx, "DELETE", tenantPath(tenant, "cases", id, "actions", namespace, flowId), nil, nil)
}

// RunCaseAction runs one of a case's attached flow actions, returning the created execution id.
func (a *CasesAPI) RunCaseAction(ctx context.Context, tenant, id string, request CaseRunActionRequest) (map[string]string, error) {
	return doJSON[map[string]string](&a.baseAPI, ctx, "POST", tenantPath(tenant, "cases", id, "actions", "run"), request, nil)
}

File: python/python-sdk/kestrapy/api/cases_api.py
import os
from typing import Any, Dict, List, Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.bulk_response import BulkResponse
from kestrapy.models.cases_controller_api_case import CasesControllerApiCase
from kestrapy.models.cases_controller_api_case_event import CasesControllerApiCaseEvent
from kestrapy.models.paged_results_cases_controller_api_case import (
    PagedResultsCasesControllerApiCase,
)
from kestrapy.models.paged_results_cases_controller_api_case_asset import (
    PagedResultsCasesControllerApiCaseAsset,
)
from kestrapy.models.paged_results_cases_controller_api_case_event import (
    PagedResultsCasesControllerApiCaseEvent,
)
from kestrapy.models.paged_results_cases_controller_api_case_execution import (
    PagedResultsCasesControllerApiCaseExecution,
)
from kestrapy.models.paged_results_cases_controller_api_case_summary import (
    PagedResultsCasesControllerApiCaseSummary,
)
from kestrapy.models.paged_results_cases_controller_api_subject_ref import (
    PagedResultsCasesControllerApiSubjectRef,
)
from kestrapy.models.query_filter import QueryFilter


class CasesApi(BaseApi):
    """Cases (`/api/v1/{tenant}/cases`).

    Cases group related executions, assets and actions under an
    acknowledge/resolve lifecycle. The controller is an EE feature gated by the
    ``CASE`` resource, so every route answers 403 when the feature is disabled.

    Case request/response bodies are large and nested; this wrapper types the
    response envelopes (paged results, ``Case``) but accepts request bodies as
    plain ``Dict[str, Any]`` so callers can send exactly what the backend
    expects without a brittle hand-written request model.
    """

    # ---- creation from a task ----

    def create_case_from_task(self, tenant: str, body: Dict[str, Any]) -> Dict[str, Any]:
        path = self._tenant_path(tenant, "cases", "from-task")
        return self._raw_json_request("POST", path, body=body)

    # ---- search / counts ----

    def search_cases(
        self,
        tenant: str,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
        date_filter: Optional[str] = None,
    ) -> PagedResultsCasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", "search")
        params = list(self._build_query_params(page=page, size=size, dateFilter=date_filter).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._json_request("GET", path, PagedResultsCasesControllerApiCase, params=params)

    def case_counts(
        self,
        tenant: str,
        filters: Optional[List[QueryFilter]] = None,
        date_filter: Optional[str] = None,
    ) -> Dict[str, Any]:
        path = self._tenant_path(tenant, "cases", "counts")
        params = list(self._build_query_params(dateFilter=date_filter).items())
        self._append_filter_params(params, filters)
        return self._raw_json_request("GET", path, params=params)

    def case_assignees(
        self,
        tenant: str,
        filters: Optional[List[QueryFilter]] = None,
        date_filter: Optional[str] = None,
    ) -> PagedResultsCasesControllerApiSubjectRef:
        path = self._tenant_path(tenant, "cases", "assignees")
        params = list(self._build_query_params(dateFilter=date_filter).items())
        self._append_filter_params(params, filters)
        return self._json_request("GET", path, PagedResultsCasesControllerApiSubjectRef, params=params)

    # ---- CRUD ----

    def create_case(self, tenant: str, body: Dict[str, Any]) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def get_case(
        self, tenant: str, id: str, allow_deleted: Optional[bool] = None
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id)
        params = self._build_query_params(allowDeleted=allow_deleted)
        return self._json_request("GET", path, CasesControllerApiCase, params=params)

    def update_case(self, tenant: str, id: str, body: Dict[str, Any]) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id)
        return self._json_request("PUT", path, CasesControllerApiCase, body=body)

    def delete_case(self, tenant: str, id: str) -> None:
        path = self._tenant_path(tenant, "cases", id)
        self._void_request("DELETE", path)

    def delete_cases_by_ids(self, tenant: str, case_ids: List[str]) -> BulkResponse:
        path = self._tenant_path(tenant, "cases", "by-ids", "delete")
        return self._json_request("POST", path, BulkResponse, body=case_ids)

    def delete_cases_by_query(
        self,
        tenant: str,
        filters: Optional[List[QueryFilter]] = None,
        date_filter: Optional[str] = None,
    ) -> BulkResponse:
        path = self._tenant_path(tenant, "cases", "by-query")
        params = list(self._build_query_params(dateFilter=date_filter).items())
        self._append_filter_params(params, filters)
        return self._json_request("DELETE", path, BulkResponse, params=params)

    # ---- lifecycle ----

    def acknowledge_cases_by_ids(self, tenant: str, case_ids: List[str]) -> BulkResponse:
        path = self._tenant_path(tenant, "cases", "by-ids", "acknowledge")
        return self._json_request("POST", path, BulkResponse, body=case_ids)

    def acknowledge_case(self, tenant: str, id: str) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "acknowledge")
        return self._json_request("POST", path, CasesControllerApiCase)

    def resolve_case(self, tenant: str, id: str, body: Dict[str, Any]) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "resolve")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def cancel_case(
        self, tenant: str, id: str, body: Optional[Dict[str, Any]] = None
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "cancel")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def change_case_status(
        self, tenant: str, id: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "status")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def assign_case(self, tenant: str, id: str, body: Dict[str, Any]) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "assign")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def follow_case(self, tenant: str, id: str) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "follow")
        return self._json_request("POST", path, CasesControllerApiCase)

    def unfollow_case(self, tenant: str, id: str) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "unfollow")
        return self._json_request("POST", path, CasesControllerApiCase)

    # ---- events / comments / attachments ----

    def case_events(
        self, tenant: str, id: str, page: Optional[int] = None, size: Optional[int] = None
    ) -> PagedResultsCasesControllerApiCaseEvent:
        path = self._tenant_path(tenant, "cases", id, "events")
        params = self._build_query_params(page=page, size=size)
        return self._json_request("GET", path, PagedResultsCasesControllerApiCaseEvent, params=params)

    def add_case_comment(
        self,
        tenant: str,
        id: str,
        body: Optional[str] = None,
        file_path: Optional[str] = None,
    ) -> CasesControllerApiCaseEvent:
        """Post a comment on a case with an optional markdown body and an optional
        file attachment. Sent as ``multipart/form-data`` (fields ``body`` and
        ``files``)."""
        path = self._tenant_path(tenant, "cases", id, "comments")
        files: Dict[str, Any] = {}
        if body is not None:
            files["body"] = (None, body)
        if file_path is not None:
            with open(file_path, "rb") as f:
                files["files"] = (os.path.basename(file_path), f.read())
        resp = self._request("POST", path, files=files)
        return self._deserialize(resp.json(), CasesControllerApiCaseEvent)

    def download_case_attachment(self, tenant: str, id: str, attachment_id: str) -> bytes:
        path = self._tenant_path(tenant, "cases", id, "attachments", attachment_id)
        return self._download_request("GET", path)

    # ---- linked executions ----

    def case_executions(
        self, tenant: str, id: str, page: Optional[int] = None, size: Optional[int] = None
    ) -> PagedResultsCasesControllerApiCaseExecution:
        path = self._tenant_path(tenant, "cases", id, "executions")
        params = self._build_query_params(page=page, size=size)
        return self._json_request(
            "GET", path, PagedResultsCasesControllerApiCaseExecution, params=params
        )

    def link_case_executions(
        self, tenant: str, id: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "executions")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def link_case_executions_by_query(
        self, tenant: str, id: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "executions", "by-query")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def unlink_case_execution(self, tenant: str, id: str, execution_id: str) -> None:
        path = self._tenant_path(tenant, "cases", id, "executions", execution_id)
        self._void_request("DELETE", path)

    # ---- auto-attach ----

    def enable_case_auto_attach(
        self, tenant: str, id: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "auto-attach")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def disable_case_auto_attach(self, tenant: str, id: str) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "auto-attach")
        return self._json_request("DELETE", path, CasesControllerApiCase)

    # ---- creation from executions ----

    def create_case_from_executions(
        self, tenant: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", "from-executions")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def create_case_from_executions_by_query(
        self, tenant: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", "from-executions", "by-query")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def cases_by_executions(self, tenant: str, execution_ids: List[str]) -> Dict[str, Any]:
        path = self._tenant_path(tenant, "cases", "by-executions")
        return self._raw_json_request("POST", path, body=execution_ids)

    # ---- assets ----

    def case_assets(self, tenant: str, id: str) -> PagedResultsCasesControllerApiCaseAsset:
        path = self._tenant_path(tenant, "cases", id, "assets")
        return self._json_request("GET", path, PagedResultsCasesControllerApiCaseAsset)

    def attach_case_asset(
        self, tenant: str, id: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "assets")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def detach_case_asset(self, tenant: str, id: str, asset_id: str) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "assets", asset_id)
        return self._json_request("DELETE", path, CasesControllerApiCase)

    def cases_by_asset(self, tenant: str, asset_id: str) -> PagedResultsCasesControllerApiCaseSummary:
        path = self._tenant_path(tenant, "cases", "by-asset", asset_id)
        return self._json_request("GET", path, PagedResultsCasesControllerApiCaseSummary)

    # ---- actions ----

    def attach_case_action(
        self, tenant: str, id: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "actions")
        return self._json_request("POST", path, CasesControllerApiCase, body=body)

    def update_case_action(
        self, tenant: str, id: str, namespace: str, flow_id: str, body: Dict[str, Any]
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "actions", namespace, flow_id)
        return self._json_request("PUT", path, CasesControllerApiCase, body=body)

    def detach_case_action(
        self, tenant: str, id: str, namespace: str, flow_id: str
    ) -> CasesControllerApiCase:
        path = self._tenant_path(tenant, "cases", id, "actions", namespace, flow_id)
        return self._json_request("DELETE", path, CasesControllerApiCase)

    def run_case_action(self, tenant: str, id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        path = self._tenant_path(tenant, "cases", id, "actions", "run")
        return self._raw_json_request("POST", path, body=body)
